### ISSUE TITLE
Use metering for P2P shuffling instrumentation

### ISSUE THREAD

**Issue by fjetter** [dask/distributed#7943]
The P2P extensions (primarily the buffers) are implementing their own version for diagnostics, e.g. here https://github.com/dask/distributed/blob/429ef8cc682da36017a28e061d1c773066470fe3/distributed/shuffle/_buffer.py#L267-L272

I think P2P diagnostics would greatly benefit from the metering / fine grained metrics and we should consider replacing (os extending) the custom ctx managers there with the metrics.meter (mind that there is also a special dashboard for P2P shuffling that is using this custom diagnostics. I don't want to break this dashboard).

cc @crusaderky @hendrikmakait

**Comment by crusaderky** [comment]
I'm delivering this ticket in 4 incremental PRs for ease of review.
Each PR incorporates the previous ones. While they make sense on their own, I would encourage reviewing them holistically.

- #8364
This adds Fine Performance Metrics to everything except the background tasks. Everything is logged under 'execute' as normal for tasks and is visible from both the bokeh dashboard as well as from the coiled dashboard.
This already shows interesting data - e.g. how much it takes to break each partition into shards and serialize them. Unsurprisingly, you will observe that most of the time is spent waiting on comms `ResourceLimiter.wait_for_available` method - in other words, waiting for the background task to free up.
- #8365 
This adds Fine Performance Metrics to the background tasks, but they end up lost in the void as there are no callbacks capturing them.
- #8366 
This captures the Fine Performance Metrics for the background tasks and attributes them to spans. They are stored under a new metrics context tag (the first element of the tuple), `shuffle`, that goes alongside `execute`, `gather-dep`, `get-data` and `memory-monitor`.
Additionally, this PR duplicates all _foreground_ metrics so that they appear both under `shuffle` and under `execute`. This is ease analysis. Metrics are separated into sub-sections 
  - `(shuffle, foreground, ...)`
  - `(shuffle, background-comms, ...)`
  - `(shuffle, background-comms, ...)`
- #8367 
This PR finally cleans up all the bespoke shuffle diagnostics, which should now be redundant.

A/B test shows no performance impact.

This introduces a minor regression in the Coiled dashboard: 
- https://github.com/coiled/platform/issues/3898

# Out of scope
- Visualize the new metrics on the bokeh dashboard
- Visualize the new metrics on the coiled dashboard (IMHO we shouldn't; this is a nitty gritty internal thing)
- Extract insights from coiled/benchmarks (everything should be recorded on the Coiled database; cluster IDs are available here https://github.com/coiled/benchmarks/actions/runs/6917401769 in benchmarks.db)
